src/storage/src/hummock/compactor/mod.rs

```diff
@@ -608,3 +614,205 @@ pub fn start_compactor(
 
     (join_handle, shutdown_tx)
 }
+
+/// The background compaction thread that receives compaction tasks from hummock compaction
+/// manager and runs compaction tasks.
+#[cfg_attr(coverage, no_coverage)]
+pub fn start_shared_compactor(
+    grpc_proxy_client: GrpcCompactorProxyClient,
+    mut receiver: mpsc::UnboundedReceiver<Request<DispatchCompactionTaskRequest>>,
+    context: CompactorContext,
+) -> (JoinHandle<()>, Sender<()>) {
+    type CompactionShutdownMap = Arc<Mutex<HashMap<u64, Sender<()>>>>;
+    let task_progress = context.task_progress_manager.clone();
+    let (shutdown_tx, mut shutdown_rx) = tokio::sync::oneshot::channel();
+    let periodic_event_update_interval = Duration::from_millis(1000);
+
+    let join_handle = tokio::spawn(async move {
+        let shutdown_map = CompactionShutdownMap::default();
+
+        let mut periodic_event_interval = tokio::time::interval(periodic_event_update_interval);
+        let executor = context.compaction_executor.clone();
+        let mut report_heartbeat_client = grpc_proxy_client.clone();
+        'consume_stream: loop {
+            let request: Option<Request<DispatchCompactionTaskRequest>> = tokio::select! {
+                _ = periodic_event_interval.tick() => {
+                    let progress_list = get_task_progress(task_progress.clone());
+                    let report_compaction_task_request = ReportCompactionTaskRequest{
+                        event: Some(ReportCompactionTaskEvent::HeartBeat(
+                            SharedHeartBeat {
+                                progress: progress_list
+                            }
+                        )),
+                     };
+                    if let Err(e) = report_heartbeat_client.report_compaction_task(report_compaction_task_request).await{
+                        tracing::warn!("Failed to report heartbeat {:#?}", e);
+                    }
+                    continue
+                }
+
+
+                _ = &mut shutdown_rx => {
+                    tracing::info!("Compactor is shutting down");
+                    return
+                }
+
+                request = receiver.recv() => {
+                    request
+                }
+
+            };
+            match request {
+                Some(request) => {
+                    let context = context.clone();
+                    let shutdown = shutdown_map.clone();
+
+                    let mut report_task_client = grpc_proxy_client.clone();
+                    let get_sst_id_client = grpc_proxy_client.clone();
```

Comment:
The previous issue is not entirely resolved. We only need to clone once and don't need to hold two clients here.